### PR TITLE
Avoid table-scanning users at startup

### PR DESCRIPTION
**Pull request by richvdh**
This takes about 10 seconds in the best case; often more.

### CODE REVIEW

File: synapse/storage/databases/main/__init__.py [hunk 1]
```diff
@@ -591,22 +592,24 @@ def check_database_before_upgrade(cur, database_engine, config: HomeServerConfig
...
-    if num_not_matching == 0:
+    # if there are any users in the database, check that the username matches our
+    # configured server name.
+
+    cur.execute("SELECT name FROM users LIMIT 1")
+    rows = cur.fetchall()
+    if not rows:
+        return
+
+    mxid = UserID.from_string(rows[0][0])
```
Comment (erikjohnston) [inline synapse/storage/databases/main/__init__.py:605]: I'd probably mildly prefer `get_domain_from_id` here

File: changelog.d/8271.bugfix [hunk 2]
```diff
@@ -0,0 +1 @@
+Remove slow table scan of `users` table from startup code.
```
Comment (erikjohnston) [inline changelog.d/8271.bugfix:1]:
```suggestion
Fix slow start times for large servers by removing a table scan of the `users` table from startup code.
```

**Review by erikjohnston — APPROVED** [review]

File: synapse/storage/databases/main/__init__.py (hunk 1 above)
Comment (richvdh) [reply synapse/storage/databases/main/__init__.py]: done

File: synapse/storage/databases/main/__init__.py [hunk 3]
```diff
@@ -591,22 +592,24 @@ def check_database_before_upgrade(cur, database_engine, config: HomeServerConfig
...
+    if not rows:
+        return
+
+    user_domain = get_domain_from_id(rows[0][0])
+    if user_domain == config.server_name:
         return
 
     raise Exception(
         "Found users in database not native to %s!\n"
         "You cannot changed a synapse server_name after it's been configured"
```
Comment (babolivier) [inline synapse/storage/databases/main/__init__.py:611]:
```suggestion
        "You cannot change a synapse server_name after it's been configured"
```

Probably worth fixing this typo while we're here?

**Pull request merged**